ControlWork/Queue.cs

```diff
@@ -0,0 +1,67 @@
+﻿namespace ControlWork;
+
+public class ThreadSafetyQueue<T, P>
+    where P : IComparable
+{
+    private readonly List<QueueElement<T, P>> _elements = new();
+    private readonly object _locker = new();
+    private int _size = 0;
+
+    public void Enqueue(T element, P priority)
+    {
+        lock (_locker)
+        {
+            _elements.Add(new QueueElement<T, P>(element, priority));
+        }
+
+        Interlocked.Increment(ref _size);
+        Thread.Sleep(100);
+    }
+
+    public T Dequeue()
+    {
+        lock (_locker)
+        {
+            if (_elements.Count == 0)
+            {
+                Monitor.PulseAll(_locker);
+                // каждые 100 мс проверяем, есть ли что-нибудь в очереди
+                while (true)
+                {
+                    if (_elements.Count > 0)
+                    {
+                        break;
+                    }
+
+                    Thread.Sleep(100);
+                }
+
+                Monitor.Wait(_locker);
+            }
```

Comment:
```suggestion
            while (_elements.Count == 0)
            {
                Monitor.Wait(_locker);
            }
```
deadlock.
Если в очереди пусто, и какой-либо поток захочет взять элемент, то он заблокирует возможность вставки.
Необходимо, чтобы ожидающий поток находился в ожидании и "отпустил" объект, по которому происходит блокировка.

Reply:
-1